Mathlib/Geometry/Manifold/VectorBundle/GramSchmidtOrtho.lean

```diff
@@ -0,0 +1,299 @@
+    ∀ i₀, gramSchmidt s i₀ x = s i₀ x := by
+  intro i
+  rw [gramSchmidt_def]
+  trans s i x - 0
+  · congr
+    apply Finset.sum_eq_zero
+    intro j hj
+    rw [Submodule.coe_eq_zero]
+    suffices span ℝ ((s · x) '' Set.Iic j) ⟂ ℝ ∙ s i x by
+      apply orthogonalProjection_mem_subspace_orthogonalComplement_eq_zero
+      rw [mem_orthogonal_singleton_iff_inner_left, ← mem_orthogonal_singleton_iff_inner_right]
+      exact this <| gramSchmidt_mem_span _ _ le_rfl
+    rw [isOrtho_span]
+    rintro u ⟨k, hk, rfl⟩ v (rfl : v = s i x)
+    apply hs
+    exact (lt_of_le_of_lt hk (Finset.mem_Iio.mp hj)).ne
+  · simp
+
+theorem gramSchmidt_ne_zero_coe (n : ι)
+    (h₀ : LinearIndependent ℝ ((s · x) ∘ ((↑) : Set.Iic n → ι))) : gramSchmidt s n x ≠ 0 :=
+  InnerProductSpace.gramSchmidt_ne_zero_coe _ h₀
+
+variable (s) in
+/-- If the input sections of `gramSchmidt` are point-wise linearly independent,
+the resulting sections are non-zero. -/
+theorem gramSchmidt_ne_zero (n : ι) (h₀ : LinearIndependent ℝ (s · x)) :
+    gramSchmidt s n x ≠ 0 :=
+  InnerProductSpace.gramSchmidt_ne_zero _ h₀
+
+-- No version of `gramSchmidt_triangular` at the moment, for technical reasons: it would expect a
+-- `Basis` (of vectors in `E x`) as input, whereas we would want a hypothesis "the section values
+-- `s i x` form a basis" instead.
+
+/-- `gramSchmidt` produces point-wise linearly independent sections when given linearly
+independent sections. -/
+theorem gramSchmidt_linearIndependent (h₀ : LinearIndependent ℝ (s · x)) :
+    LinearIndependent ℝ (gramSchmidt s · x) :=
+  InnerProductSpace.gramSchmidt_linearIndependent h₀
+
+/-- When the sections `s` form a basis at `x`, so do the sections `gramSchmidt s`. -/
+noncomputable def gramSchmidtBasis (hs : LinearIndependent ℝ (s · x))
+    (hs' : ⊤ ≤ Submodule.span ℝ (Set.range (s · x))) :
+    Basis ι ℝ (E x) :=
+  Basis.mk (gramSchmidt_linearIndependent hs)
+    ((span_gramSchmidt s x).trans (eq_top_iff'.mpr fun _ ↦ hs' trivial)).ge
+
+theorem coe_gramSchmidtBasis (hs : LinearIndependent ℝ (s · x))
+    (hs' : ⊤ ≤ Submodule.span ℝ (Set.range (s · x))) :
+    (gramSchmidtBasis hs hs') = (gramSchmidt s · x) :=
+  Basis.coe_mk _ _
+
+/-- The normalized `gramSchmidt`, i.e. each resulting section has unit length (or vanishes)
+at each point -/
+noncomputable def gramSchmidtNormed
+    (s : ι → (x : B) → E x) (n : ι) : (x : B) → E x := fun x ↦
+  InnerProductSpace.gramSchmidtNormed ℝ (s · x) n
+
+lemma gramSchmidtNormed_coe {n : ι} :
+    gramSchmidtNormed s n x = ‖gramSchmidt s n x‖⁻¹ • gramSchmidt s n x := by
+  simp [gramSchmidtNormed, InnerProductSpace.gramSchmidtNormed]
+
+theorem gramSchmidtNormed_unit_length_coe {n : ι}
+    (h₀ : LinearIndependent ℝ ((s · x) ∘ ((↑) : Set.Iic n → ι))) :
+    ‖gramSchmidtNormed s n x‖ = 1 :=
+  InnerProductSpace.gramSchmidtNormed_unit_length_coe n h₀
+
+-- XXX: should n be explicit or implicit? (same question in InnerProductSpace)
```

Comment:
@faenuccio What do you think about this one?

Reply:
IMHO it should be implicit both here and there. Since there is no condition on `n`, even if one were to use this as a `rw` lemma, a `rw gSNormed_unit_length` would do, because `n` would be picked up automatically (the only problem would arise if one wants to `rw [← ...]` to change a `1` into a norm, but this will probably never happen).